drivers/spi/spi_ifx_cat1.c

```diff
@@ -0,0 +1,356 @@
+#include <zephyr/drivers/spi.h>
+#include <zephyr/kernel.h>
+
+#include <cyhal_scb_common.h>
+#include <cyhal_spi.h>
+
+#define IFX_CAT1_SPI_LOCK_TMOUT_MS      (30 * 1000)
+#define IFX_CAT1_SPI_DEFAULT_OVERSAMPLE (4)
+#define IFX_CAT1_SPI_MIN_DATA_WIDTH     (8)
+#define IFX_CAT1_SPI_MAX_DATA_WIDTH     (32)
+
+/* Device config structure */
+struct ifx_cat1_spi_config {
+	CySCB_Type *reg_addr;
+	const struct pinctrl_dev_config *pcfg;
+	cy_stc_scb_spi_config_t scb_spi_config;
+	uint8_t irq_priority;
+};
+
+/* Data structure */
+struct ifx_cat1_spi_data {
+	struct spi_context ctx;
+	cyhal_spi_t obj; /* SPI CYHAL object */
+	cyhal_resource_inst_t hw_resource;
+	struct spi_config spi_cfg_cache;
+	size_t chunk_len;
+};
+
+static int32_t _get_hw_block_num(CySCB_Type *reg_addr)
+{
+	uint32_t i;
+
+	for (i = 0u; i < _SCB_ARRAY_SIZE; i++) {
+		if (_CYHAL_SCB_BASE_ADDRESSES[i] == reg_addr) {
+			return i;
+		}
+	}
+
+	return -ENOMEM;
+}
+
+static uint8_t _get_dfs_value(struct spi_context *ctx)
+{
+	switch (SPI_WORD_SIZE_GET(ctx->config->operation)) {
+	case 8:
+		return 1;
+	case 16:
+		return 2;
+	case 32:
+		return 4;
+	default:
+		return 1;
+	}
+}
+
+static void _transfer_chunk(const struct device *dev)
+{
+	struct ifx_cat1_spi_data *const data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	int ret = 0;
+	size_t chunk_len = spi_context_max_continuous_chunk(ctx);
+
+	if (chunk_len == 0) {
+		goto exit;
+	}
+
+	data->chunk_len = chunk_len;
+
+	cy_rslt_t result = cyhal_spi_transfer_async(
+		&data->obj, ctx->tx_buf, spi_context_tx_buf_on(ctx) ? chunk_len : 0, ctx->rx_buf,
+		spi_context_rx_buf_on(ctx) ? chunk_len : 0);
+	if (result == CY_RSLT_SUCCESS) {
+		return;
+	}
+
+	ret = -EIO;
+
+exit:
+	spi_context_cs_control(ctx, false);
+	spi_context_complete(ctx, dev, ret);
+}
+
+static void _spi_interrupt_callback(void *arg, cyhal_spi_event_t event)
+{
+	const struct device *dev = (const struct device *)arg;
+	struct ifx_cat1_spi_data *const data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+
+	if (event & CYHAL_SPI_IRQ_ERROR) {
+#if defined(CONFIG_SPI_ASYNC)
+		cyhal_spi_abort_async(&data->obj);
+#endif
+		spi_context_cs_control(ctx, false);
+		spi_context_complete(ctx, dev, -EIO);
+	}
+
+	if (event & CYHAL_SPI_IRQ_DONE) {
+		spi_context_update_tx(ctx, _get_dfs_value(ctx), data->chunk_len);
+		spi_context_update_rx(ctx, _get_dfs_value(ctx), data->chunk_len);
+
+		_transfer_chunk(dev);
+	}
+}
+
+int _spi_config(const struct device *dev, const struct spi_config *spi_cfg)
+{
+	cy_rslt_t result;
+	struct ifx_cat1_spi_data *const data = dev->data;
+	const struct ifx_cat1_spi_config *const config = dev->config;
+	cy_stc_scb_spi_config_t scb_spi_config = config->scb_spi_config;
+	struct spi_context *ctx = &data->ctx;
+	bool spi_mode_cpol = false;
+	bool spi_mode_cpha = false;
+
+	/* check if configuration was changed from previous run, if so skip setup again */
+	if (memcmp(&data->spi_cfg_cache, spi_cfg, sizeof(struct spi_config)) == 0) {
```

Comment:
Though the api does not define spi_config structure to be strictly constant, it is not supposed to change internally (it most of the time, if not all the time, comes from dts). You can compare the pointer directly here. 

That said, context already have a pointer for it. So no need of spi_cfg_cache then?